### ISSUE TITLE
PyRanges constructor changes input df dtypes

### ISSUE THREAD

**Issue by stephenkraemer** [biocore-ntnu/pyranges#6]
The current behavior of the PyRanges constructor when initialising from a DataFrame is to change the dtypes of the Chromosome, Start and End column of the input dataframe *inplace*.

For me this behavior has been problematic. Would consider changing it? The culprit is the set_dtypes call at the beginning of PyRanges._init.

Example:

```{python}
import pyranges as pr
import pandas as pd

df = pd.DataFrame({
    "Chromosome": list("1122"),
    "Start": [1, 2, 3, 4],
    "End": [10, 11, 12, 13]}
)
df.dtypes
# object, i8, i8

gr = pr.PyRanges(df)

df.dtypes
# category, i4, i4
```

Btw: thanks again for your great work!

**Comment by endrebak** [comment]
Yes, this has been nagging me that it changes stuff inplace, implicitly. I have a flag `copy_df`, but it is more verbose than just calling copy on the dataframe. I want PyRanges to be able to steal the dataframe to save memory. To avoid this behavior use

`gr = pr.PyRanges(df.copy())`

There is a second problem here: 

I think most people most of the time want automatic detection and setting of appropriate data types, but this behavior could also be annoying. Perhaps I should have a flag `dtypes` where “auto” auto-detects, False does nothing and a dict does the same as 🐼.

**Comment by endrebak** [comment]
Thanks for creating issues. It helps me understand where the pain points are. I’ve only just started using pyranges myself, so I guess I will find and fix thousands of these small issues myself over the coming months :)

**Comment by endrebak** [comment]
Actually, I think copying the df should be the default behavior, and that there should be a copy_df flag you can set to False if you want to steal it. Will fix within a week.

**Comment by stephenkraemer** [comment]
Great, thanks for taking this into consideration! I think this is indeed more what people expect, also because they are used to it from the the defensive copying in pandas.

**Comment by endrebak** [comment]
Should be fixed in 0.0.38. I am going to be very conservative with changes, but in this case I was just plain wrong (much of my testing/early use was on "too large for GRanges"-data and in that particular instance copying took very much longer).
